File: templates/python-fastapi/src/__package__/domain/exceptions.py
from __future__ import annotations


class DomainError(Exception):
    """Base class for all domain-level errors."""


class UserNotFoundError(DomainError):
    def __init__(self, user_id: int) -> None:
        super().__init__(f"User not found: {user_id}")
        self.user_id = user_id


class EmailAlreadyTakenError(DomainError):
    def __init__(self, email: str) -> None:
        super().__init__(f"Email already taken: {email}")
        self.email = email

File: templates/ts-nestjs/src/modules/users/domain/exceptions.ts
export class UserNotFoundException extends Error {
  constructor(id: string) {
    super(`User not found: ${id}`);
    this.name = 'UserNotFoundException';
  }
}

export class EmailAlreadyTakenException extends Error {
  constructor(email: string) {
    super(`Email already taken: ${email}`);
    this.name = 'EmailAlreadyTakenException';
  }
}
